### ISSUE TITLE
Networks with certain names prevent connecting to containers

### ISSUE THREAD

**Issue by yousefamar** [docker/docker#23346]
**Output of `docker version`:**

```
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.6.2
 Git commit:   5604cbe
 Built:        Mon May  2 00:06:51 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.6.2
 Git commit:   5604cbe
 Built:        Mon May  2 00:06:51 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 277
Server Version: 1.11.1
Storage Driver: devicemapper
 Pool Name: docker-8:6-155885-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 4.859 GB
 Data Space Total: 107.4 GB
 Data Space Available: 10.33 GB
 Metadata Space Used: 13.85 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.134 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.122 (2016-04-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.5.4-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.741 GiB
Name: amarmac
ID: 5PE5:IJQC:N6CV:KS47:TCGF:IX3K:AKBA:6OKX:RLEW:CDAA:LNNJ:CQ62
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
```
~ » uname -a
Linux amarmac 4.5.4-1-ARCH #1 SMP PREEMPT Wed May 11 22:21:28 CEST 2016 x86_64 GNU/Linux
```

**Steps to reproduce the issue:**

1. Launch a Docker container that exposes and publishes a port, but without specifying a host port. For example, this hello-world web server:

    ```
    docker run -d -p 80 --name c1 tutum/hello-world
    ```

2. Get the port, and browse to it in your favourite web browser (or curl, or whatever). Notice that everything works just fine.

    ```
    docker port c1 80
    ```

3. Create a Docker network called "a" (or anything starting with an a) and connect your container to it. Notice that now connections to the container are refused.

    ```
    docker network create a
    docker network connect a c1
    ```

4. Repeat 1-3 but this time with a container name that does not start with "a". Notice how connections are not refused.
5. Proceed to tear your hair out after days of debugging.

**Describe the results you received:**

Host connections to containers connected to a Docker network with a name that starts with "a" are refused, but not for ones named anything else.

**Describe the results you expected:**

Host connections to containers connected to a Docker network with any name are *not* refused.

**Additional information you deem important (e.g. issue happens only occasionally):**

- The steps for reproduction use the Docker CLI, but I can confirm that the issue persists when using the Docker remote API.
- This only happens when you do not specify a host port (i.e. you let Docker map to some ephemeral port).
- This always happens when the name of the network begins with an "a", though it could happen for other names too; I haven't tested that many names.

**Comment by thaJeztah** [comment]
Perhaps this may be fixed by https://github.com/docker/docker/pull/23011 for some cases, but I guess this is due to the order in which the container is connected, and the port is published on the first network (alphabetically). Also, the random port may change on network connect, which should be fixed in docker 1.11.2 (see https://github.com/docker/docker/pull/22997, https://github.com/docker/docker/issues/22836)

**Comment by thaJeztah** [comment]
I think there's an existing issue about this, but I should look it up

**Comment by yousefamar** [comment]
Thanks for the reply! I've just tested this with 1.11.2 and the problem persists.

You're absolutely right though - I can confirm that it happens when the network is named "bridgd" (< bridge), but not when the network is named "bridgf" (> bridge). I don't think #23011 fixes it (maybe that's only for container links?).

Do you know if there's a way to control which network the port is published on (or lock it on bridge)? Could you link me to that part of the docs if so?

**Comment by thaJeztah** [comment]
Perhaps it helps to create the custom networks with `--internal`, so that the port is not published on that network. The user guide contains a section on container networking; https://docs.docker.com/engine/userguide/networking/dockernetworks/

**Comment by yousefamar** [comment]
Thanks for the tip @thaJeztah! For posterity: making the networks internal doesn't seem to have an effect, but at least for now I can make sure that my network names are just > bridge.

**Comment by aboch** [comment]
Fixed by #22997
